x/lazyledgerapp/types/payformessage.go

```diff
@@ -0,0 +1,288 @@
+package types
+
+import (
+	"crypto/sha256"
+	"errors"
+	fmt "fmt"
+
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lazyledger/lazyledger-core/crypto/merkle"
+	"github.com/lazyledger/nmt"
+)
+
+const (
+	TypeMsgPayforMessage                   = "payformessage"
+	TypeSignedTransactionDataPayForMessage = "signedtransactiondatapayformessage"
+	ShareSize                              = 256
+)
+
+///////////////////////////////////////
+// 	MsgWirePayForMessage
+///////////////////////////////////////
+
+var _ sdk.MsgRequest = &MsgWirePayForMessage{}
+
+func (msg *MsgWirePayForMessage) Route() string { return RouterKey }
+
+func (msg *MsgWirePayForMessage) Type() string { return TypeMsgPayforMessage }
+
+// ValidateBasic checks for valid namespace length, declared message size, share
+// commitments, signatures for those share commitments, and fulfills the sdk.Msg
+// interface
+func (msg *MsgWirePayForMessage) ValidateBasic() error {
+	pubK := msg.PubKey()
+
+	// ensure that the namespace id is of length == 8
+	if len(msg.GetMessageNameSpaceId()) != 8 {
+		return fmt.Errorf(
+			"invalid namespace length: got %d wanted 8",
+			len(msg.GetMessageNameSpaceId()),
+		)
+	}
+
+	// ensure that the included message is evenly divisble into shares
+	if uint64(len(msg.GetMessage()))%ShareSize != 0 {
+		return fmt.Errorf("Share message must be divisible by %d", ShareSize)
+	}
+
+	// make sure that the message size matches the actual size of the message
+	if msg.MessageSize != uint64(len(msg.Message)) {
+		return fmt.Errorf(
+			"Declared Message size does not match actual Message size, %d vs %d",
+			msg.MessageSize,
+			len(msg.Message),
+		)
+	}
+
+	for _, commit := range msg.MessageShareCommitment {
+		// check that each commit is valid
+		calculatedCommit, err := CreateCommit(commit.K, msg.GetMessageNameSpaceId(), msg.Message)
+		if err != nil {
+			return err
+		}
+
+		if string(calculatedCommit) != string(commit.ShareCommitment) {
+			return fmt.Errorf("invalid commit for square size %d", commit.K)
+		}
+
+		// check that the signatures are valid
+		bytesToSign, err := msg.GetCommitSignBytes(commit.K)
+		if err != nil {
+			return err
+		}
+
+		if !pubK.VerifySignature(bytesToSign, commit.Signature) {
+			return fmt.Errorf("invalid signature for share commitment to square size %d", commit.K)
+		}
+	}
+
+	return nil
+}
+
+// GetSignBytes returns messages bytes that need to be signed in order for the
+// message to be valid todo(evan): follow the spec so that each share commitment
+// is signed, instead of the entire message
+// TODO(evan): remove panic and add possibility to vary square size
+func (msg *MsgWirePayForMessage) GetSignBytes() []byte {
+	out, err := msg.GetCommitSignBytes(64)
```

Comment:
What does 64 stand for?

Reply:
that was the square size (width), it has been switched to using a constant like you suggested earlier. Using constants for all hard coded values makes a lot of sense, thanks for this input!